src/Features/LanguageServer/Protocol/LanguageServerProtocol.cs

```diff
@@ -1,45 +1,146 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
 using System.Threading;
 using System.Threading.Tasks;
-using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
 using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
 
 namespace Microsoft.CodeAnalysis.LanguageServer
 {
     /// <summary>
     /// Implements Language Server Protocol
+    /// TODO - Make this public when we're ready.
     /// </summary>
-    public sealed class LanguageServerProtocol
+    [Shared]
+    [Export(typeof(LanguageServerProtocol))]
+    internal sealed class LanguageServerProtocol
     {
-        // for now, it is null but later we might use this to provide info on types on dlls
-        private readonly IMetadataAsSourceFileService _metadataAsSourceService = null;
+        private readonly ImmutableDictionary<string, Lazy<IRequestHandler, IRequestHandlerMetadata>> _requestHandlers;
 
-        // TODO - Move hierarchicalDocumentSymbolSupport to client capabilities.
-        // https://github.com/dotnet/roslyn/projects/45#card-20033973
-        private readonly bool _hierarchicalDocumentSymbolSupport;
-        private readonly LSP.ClientCapabilities _clientCapabilities;
+        [ImportingConstructor]
+        public LanguageServerProtocol([ImportMany] IEnumerable<Lazy<IRequestHandler, IRequestHandlerMetadata>> requestHandlers)
+        {
+            _requestHandlers = CreateMethodToHandlerMap(requestHandlers);
+        }
+
+        private static ImmutableDictionary<string, Lazy<IRequestHandler, IRequestHandlerMetadata>> CreateMethodToHandlerMap(IEnumerable<Lazy<IRequestHandler, IRequestHandlerMetadata>> requestHandlers)
+        {
+            var requestHandlerDictionary = new Dictionary<string, Lazy<IRequestHandler, IRequestHandlerMetadata>>();
```

Comment:
you could use ImmutableDictionaryBuilder rather than do dictionary -> immutable conversion.